File: rotate_Array_from_k.js
const reverse = function(num, start, end){

    while(start<end){
        [num[start], num[end]] = [num[end], num[start]];
        start++;
        end--;
    }
}

const rotate = function(num, k){
    k = k%num.length;
    reverse(num,0, num.length-1); // full array reverse
    reverse(num, 0, k-1);// from 0 index to k-1
    reverse(num, k, num.length-1);

    return num;
}

console.log(rotate([1, 2, 3, 4, 5], 2));

File: rotate_array_from_k.py
# Brute force solution time complexity = O(k*n) space = O(k*n)

def rotate_array_from_k(arr, k):
    if k == 0:
        return arr
    else:
        return rotate_array_from_k(arr[1:] + [arr[0]], k-1)
    
print(rotate_array_from_k([1,2,3],1))


#Efficient 

def reverse(arr, start, end):
    while(start<end):
        arr[start], arr[end] = arr[end], arr[start]
        start += 1
        end -= 1
        
        
def rotate_array(arr, k):
    n = len(arr)
    k = k%n
    
    reverse(arr, 0, n - 1)
    reverse(arr, 0, k-1)
    reverse(arr, k, n-1)
    
    return arr

arr = [1,2,3,4,5]
k = 2

rotated_array = rotate_array(arr, k)
print(rotated_array)

